source/common/upstream/health_discovery_service.cc

```diff
@@ -298,22 +374,162 @@ HdsCluster::HdsCluster(Server::Admin& admin, Runtime::Loader& runtime,
+
+  for (auto& health_check : health_checks) {
+    uint64_t health_check_hash = MessageUtil::hash(health_check);
+
+    // Check to see if this exact same health_check config already has a health checker.
+    auto health_checker = health_checkers_map_.find(health_check_hash);
+    if (health_checker != health_checkers_map_.end()) {
+      // If it does, use it.
+      health_checkers_map.insert({health_check_hash, health_checker->second});
+      health_checkers.push_back(health_checker->second);
+    } else {
+      // If it does not, create a new one.
+      auto new_health_checker =
+          Upstream::HealthCheckerFactory::create(health_check, *this, runtime, random, dispatcher,
+                                                 access_log_manager, validation_visitor_, api);
+      health_checkers_map.insert({health_check_hash, new_health_checker});
+      health_checkers.push_back(new_health_checker);
+
+      // Start these health checks now because upstream assumes they already have been started.
+      new_health_checker->start();
+    }
+  }
+
+  // replace our member data structures with our newly created ones.
+  health_checkers_ = std::move(health_checkers);
+  health_checkers_map_ = std::move(health_checkers_map);
+}
+void HdsCluster::updateHosts(
+    const Protobuf::RepeatedPtrField<envoy::config::endpoint::v3::LocalityLbEndpoints>&
+        locality_endpoints) {
+  // Create the data structures needed for PrioritySet::update.
+  HostVectorSharedPtr hosts = std::make_shared<std::vector<HostSharedPtr>>();
+  std::vector<HostSharedPtr> hosts_added;
+  std::vector<HostSharedPtr> hosts_removed;
+  std::vector<HostVector> hosts_by_locality;
+
+  // Use for delta update comparison.
+  HostsMap hosts_map;
+
+  for (auto& endpoints : locality_endpoints) {
+    hosts_by_locality.emplace_back();
+    for (auto& endpoint : endpoints.lb_endpoints()) {
+      LocalityEndpointTuple endpoint_key = {endpoints.locality(), endpoint};
+
+      // Check to see if this exact Locality+Endpoint has been seen before.
+      auto host_pair = hosts_map_.find(endpoint_key);
+      HostSharedPtr host;
+      if (host_pair != hosts_map_.end()) {
+        // If we have this exact pair, save the shared pointer.
+        host = host_pair->second;
+      } else {
+        // We do not have this endpoint saved, so create a new one.
+        auto new_host = std::make_shared<HostImpl>(
+            info_, "", Network::Address::resolveProtoAddress(endpoint.endpoint().address()),
+            nullptr, 1, endpoints.locality(),
+            envoy::config::endpoint::v3::Endpoint::HealthCheckConfig().default_instance(), 0,
+            envoy::config::core::v3::UNKNOWN);
+
+        // Set the initial health status as in HdsCluster::initialize.
+        new_host->healthFlagSet(Host::HealthFlag::FAILED_ACTIVE_HC);
+
+        // Add to our hosts added list and save the shared pointer.
+        hosts_added.push_back(host);
```

Comment:
Should be new_host here?
Maybe no need the new_host variable, just use host to create new endpoint.

Reply:
Might worth adding an integration test (single cluster should be fine) to confirm it's working correctly

Reply:
Fixed, am now appending to hosts_added correctly. Resolving.